### DISCUSSION TITLE
What automation helper device triggers would you like to see?

### DISCUSSION THREAD

**Discussion by AdrianGarside** [greghesp/ha-bambulab#238] [Ideas]
This build adds support for a handful of device triggers that can be used in automations:
Print Canceled
Print Failed (X1 only)
Print Finished
Print Started
Printer error detected (w/ HMS error as data payload)

For the most part these are just conveniences - you could craft an automation triggering the same way but it takes a bit more effort to understand which sensor(s) to use to track the activity you want which requires understanding how the sensors behave & interact / what the available states are. The only one in the list that's not true for is the print canceled event as that's only a transient event from the printer rather than a persistent state.

These new triggers look like this in use:
![image](https://github.com/greghesp/ha-bambulab/assets/26147576/eadebf77-15c8-4919-a82a-774a49a5ba72)
(when accessed via the + automation button on the device page) and
![image](https://github.com/greghesp/ha-bambulab/assets/26147576/0cd7edd5-c33e-4ea5-b130-3b7abac0e5c0)
when you go to the automations page and add a device trigger.

So my question is - what common case scenarios could be made a little easier by adding a device trigger for it? A more complex scenario might, for example, be a dedicate 'spaghetti detected' error so you can easily trigger a notification off that specific error. Or one for the print head falling off. But there's already a generic error event so maybe specific versions isn't interesting.

So looking for suggestions on what you'd like to see present as a simplified device trigger event.

**Comment by sushibagel** [comment]
Thanks for your hard work this is great I plan to use it once I get my printer in a few days. 

A couple additional triggers I'd love to see (if possible) AMS filament change filament runout would also be cool. 

With my current printer I use Octoprint and have a notification that is sent to my phone every few layers so a layer change one would be excellent to have.

**Reply by AdrianGarside** [reply]
Filament run out is a good one, for sure.

Layer change you should be able to do do pretty trivially in home assistant directly from the layer count sensor changing. Adding a device trigger for that wouldn’t add much value.

**Comment by grkan13** [comment]
Hello,

I'm trying to create an automation by using progress level. Print start/finish triggers the automation but the progress only triggers once at the beginning. It would be nice if we can keep a trigger every single time progress level changes.

Basically, I would like to show the progress level by %5 increments via WLED

**Reply by sushibagel** [reply]
I don't have this installed yet as I'm still waiting on my printer to ship. But as @AdrianGarside said you (and this is actually how I did it for Octoprint) by creating an automation that tracks the layer number/progress using the layer sensor you should be able to do it pretty easily. 

What I did in my Octoprint one was created a helper (Z-Tracker) and setup an automation to increment the helper every time the z height changed. From there I had another automation I setup to trigger when the helper reached a certain number or higher (4 in my case), this triggered the notification and reset the helper count. 

I did also later add to it a bit to have it only send the notification every 10 minutes so I wouldn't get spammed too much on smaller prints. This was accomplished by simply turning off the second automation after it is triggered and reenabling it with a timer. 

There are likely other ways to do it but that was the way I did it.

**Reply by AdrianGarside** [reply]
Yes, this seems like something the end user should implement via a template sensor that adjusts the value to 5% increments and then you trigger an notification based on that template sensor value changing, optionally with a frequency cap as suggested above. Anything I do would necessarily have to be a single fixed behavior that may or may not be what any individual user wants so I'm trying to stay to more widely useful device automations or that are otherwise hard for the end user to implement easily.

**Reply by kanflo** [reply]
Great HASS add on! What possibilities are there to get HMS errors? I see they flow into pybambu but can they be read from HASS?

**Reply by AdrianGarside** [reply]
There's a device automation for HMS specifically: https://github.com/greghesp/ha-bambulab/blob/main/docs/DeviceTrigger.md

I'm about to improve the HMS handling by using the fact there's a severity code built into them (fatal, serious, common, info) so that can be used to easily filter notifications to only the fatal and serious ones.

**Reply by grkan13** [reply]
Appreciate the help @AdrianGarside and @sushibagel.

I've setup the sensor as you suggested. It works all good.

On Saturday, the nozzle clogged up and it ended up with a huge mess. It took almost 4 hours for me to clean up :( While I was checking the data, I saw that the print status was still printing but the layer number and progress was still at certain number for 3 hours. Maybe I can create another automation to catch an error like that by tracking those entities.
